gusto/state.py

```diff
@@ -238,31 +306,27 @@ def dump(self, t=0, diagnostic_everydump=False, pickup=False):
                 t = chk.read_attribute("/", "time")
                 next(self.dumpcount)
 
-        elif (next(self.dumpcount) % self.output.dumpfreq) == 0:
-
-            print("DBG dumping", t)
+        else:
 
             # calculate diagnostic fields
             for field in self.diagnostic_fields:
                 field(self)
 
-            # dump fields
-            self.dumpfile.write(*self.to_dump)
-
-            # dump fields on latlon mesh
-            if len(self.output.dumplist_latlon) > 0:
-                self.dumpfile_ll.write(*self.to_dump_latlon)
-
             # compute diagnostics
-            diagnostic_fns = ['min', 'max', 'rms', 'l2']
             for name in self.diagnostics.fields:
-                for fn in diagnostic_fns:
+                for fn in self.diagnostics.available_diagnostics:
                     d = getattr(self.diagnostics, fn)
                     data = d(self.field_dict[name])
-                    self.diagnostic_data[name][fn].append(data)
-                if len(self.field_dict[name].ufl_shape) is 0:
-                    data = self.diagnostics.total(self.field_dict[name])
-                    self.diagnostic_data[name]["total"].append(data)
+                    self.diagnostic_data[name][fn] = data
+            self.diagnostic_dump()
+
+            # calculate pointwise data
+            point_data = {}
+            for name, plist in self.output.point_data.iteritems():
+                # get points in the right format for the at function
+                points = [p for p in itertools.product(*plist)]
+                point_data[name] = [x.tolist() for x in self.field_dict[name].at(points)]
```

Comment:
Field evaluation is currently collective across the MPI communicator.  So you need to arrange that this iteration happens in a consistent order.  I think rather than having a dict, you should have a list of pairs `(name, points)`.  Given that you always iterate over everything.